File: r/pca.R
# PCA with SVM Classification on Wine Dataset

# Load the dataset
dataset <- read.csv('../data/wine.csv')

# Split the dataset into training and test sets
library(caTools)
set.seed(123)
split <- sample.split(dataset, SplitRatio = 0.8)
dataset_train <- subset(dataset, split == TRUE)
dataset_test  <- subset(dataset, split == FALSE)

# Feature scaling (excluding the target variable at column 14)
dataset_train[, -14] <- scale(dataset_train[, -14])
dataset_test[, -14]  <- scale(dataset_test[, -14])

# Apply PCA to reduce to 2 components
library(caret)
pca <- preProcess(dataset_train[, -14], method = 'pca', pcaComp = 2)
dataset_train <- predict(pca, dataset_train)
dataset_test  <- predict(pca, dataset_test)

# Rearranging columns: PC1, PC2, then target
dataset_train <- dataset_train[, c("PC1", "PC2", "Customer_Segment")]
dataset_test  <- dataset_test[, c("PC1", "PC2", "Customer_Segment")]

# Fit SVM classifier with a linear kernel
library(e1071)
classifier <- svm(formula = Customer_Segment ~ .,
                  data = dataset_train,
                  type = 'C-classification',
                  kernel = 'linear')

# Predict the test set
y_pred <- predict(classifier, newdata = dataset_test)

# Generate confusion matrix
cm <- table(dataset_test$Customer_Segment, y_pred)
print("Confusion Matrix:")
print(cm)

# Function to visualize results
visualize_results <- function(data, title) {
  X1 <- seq(min(data[, 1]) - 1, max(data[, 1]) + 1, by = 0.01)
  X2 <- seq(min(data[, 2]) - 1, max(data[, 2]) + 1, by = 0.01)
  grid <- expand.grid(PC1 = X1, PC2 = X2)
  y_grid <- predict(classifier, newdata = grid)
  
  plot(data[, -3],
       main = title,
       xlab = 'PC1',
       ylab = 'PC2',
       xlim = range(X1),
       ylim = range(X2))
  
  contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
  points(grid,
         pch = '.',
         col = ifelse(y_grid == 1, 'tomato',
                      ifelse(y_grid == 2, 'springgreen', 'skyblue')))
  points(data[, -3],
         pch = 21,
         bg = ifelse(data[, 3] == 1, 'red',
                     ifelse(data[, 3] == 2, 'darkgreen', 'blue')))
}

# Visualize training set
visualize_results(dataset_train, "PCA - Training Set")

# Visualize test set
visualize_results(dataset_test, "PCA - Test Set")

File: python/pca.py
# -*- coding: utf-8 -*-
"""
Principal Component Analysis (PCA) with Logistic Regression on the Wine Dataset
"""

# Import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score

# Load the dataset
dataset = pd.read_csv('../data/Wine.csv')
X = dataset.iloc[:, :-1].values  # Features
y = dataset.iloc[:, -1].values   # Target

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0
)

# Standardize the features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Apply PCA to reduce to 2 principal components
pca = PCA(n_components=2)
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)
explained_variance = pca.explained_variance_ratio_

# Train logistic regression classifier
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train, y_train)

# Predict the test set results
y_pred = classifier.predict(X_test)

# Evaluate model performance
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)

print("Confusion Matrix:\n", cm)
print("Accuracy Score:", round(accuracy, 4))

# Visualization function
def plot_decision_boundary(X_set, y_set, title):
    X1, X2 = np.meshgrid(
        np.arange(X_set[:, 0].min() - 1, X_set[:, 0].max() + 1, 0.01),
        np.arange(X_set[:, 1].min() - 1, X_set[:, 1].max() + 1, 0.01)
    )
    Z = classifier.predict(np.c_[X1.ravel(), X2.ravel()]).reshape(X1.shape)

    plt.figure(figsize=(8, 6))
    plt.contourf(X1, X2, Z, alpha=0.3, cmap=ListedColormap(('red', 'green', 'blue')))
    plt.xlim(X1.min(), X1.max())
    plt.ylim(X2.min(), X2.max())

    for i, label in enumerate(np.unique(y_set)):
        plt.scatter(
            X_set[y_set == label, 0],
            X_set[y_set == label, 1],
            c=ListedColormap(('red', 'green', 'blue'))(i),
            label=f'Class {label}'
        )

    plt.title(title)
    plt.xlabel('Principal Component 1')
    plt.ylabel('Principal Component 2')
    plt.legend()
    plt.grid(True)
    plt.tight_layout()
    plt.show()

# Visualize training set results
plot_decision_boundary(X_train, y_train, 'PCA (2D) - Wine Dataset (Training Set)')

# Visualize test set results
plot_decision_boundary(X_test, y_test, 'PCA (2D) - Wine Dataset (Test Set)')
